server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/scm/ScmInfoRepositoryImpl.java

```diff
@@ -19,102 +19,110 @@
-    this.batchReportReader = batchReportReader;
+  private final AnalysisMetadataHolder analysisMetadata;
+  private final SourceLinesDiff sourceLinesDiff;
+  private final SourceHashRepository sourceHashRepository;
+
+  public ScmInfoRepositoryImpl(BatchReportReader scannerReportReader, AnalysisMetadataHolder analysisMetadata, ScmInfoDbLoader scmInfoDbLoader,
+    SourceLinesDiff sourceLinesDiff, SourceHashRepository sourceHashRepository) {
+    this.scannerReportReader = scannerReportReader;
+    this.analysisMetadata = analysisMetadata;
     this.scmInfoDbLoader = scmInfoDbLoader;
+    this.sourceLinesDiff = sourceLinesDiff;
+    this.sourceHashRepository = sourceHashRepository;
   }
 
   @Override
-  public Optional<ScmInfo> getScmInfo(Component component) {
+  public com.google.common.base.Optional<ScmInfo> getScmInfo(Component component) {
     requireNonNull(component, "Component cannot be null");
-    return initializeScmInfoForComponent(component);
-  }
 
-  private Optional<ScmInfo> initializeScmInfoForComponent(Component component) {
     if (component.getType() != Component.Type.FILE) {
-      return Optional.absent();
-    }
-    ScmInfo scmInfo = scmInfoCache.get(component);
-    if (scmInfo != null) {
-      return optionalOf(scmInfo);
+      return com.google.common.base.Optional.absent();
     }
 
-    scmInfo = getScmInfoForComponent(component);
-    scmInfoCache.put(component, scmInfo);
-    return optionalOf(scmInfo);
+    return toGuavaOptional(scmInfoCache.computeIfAbsent(component, this::getScmInfoForComponent));
   }
 
-  private static Optional<ScmInfo> optionalOf(ScmInfo scmInfo) {
-    if (scmInfo == NoScmInfo.INSTANCE) {
-      return Optional.absent();
-    }
-    return Optional.of(scmInfo);
+  private static com.google.common.base.Optional<ScmInfo> toGuavaOptional(Optional<ScmInfo> scmInfo) {
+    return com.google.common.base.Optional.fromNullable(scmInfo.orElse(null));
   }
 
-  private ScmInfo getScmInfoForComponent(Component component) {
-    ScannerReport.Changesets changesets = batchReportReader.readChangesets(component.getReportAttributes().getRef());
-    if (changesets == null) {
-      LOGGER.trace("No SCM info for file '{}'", component.getKey());
-      return NoScmInfo.INSTANCE;
-    }
-    if (changesets.getCopyFromPrevious()) {
-      return scmInfoDbLoader.getScmInfoFromDb(component);
+  private Optional<ScmInfo> getScmInfoForComponent(Component component) {
+    ScannerReport.Changesets changesets = scannerReportReader.readChangesets(component.getReportAttributes().getRef());
+
+    if (changesets != null) {
+      if (changesets.getChangesetCount() == 0) {
+        return generateAndMergeDb(component, changesets.getCopyFromPrevious());
+      }
+      return getScmInfoFromReport(component, changesets);
     }
-    return getScmInfoFromReport(component, changesets);
+
+    LOGGER.trace("No SCM info for file '{}'", component.getKey());
+    return generateAndMergeDb(component, false);
```

Comment:
Why `false` ? If the scanner does not send any SCM information, why the DB data could not be reused ?

Reply:
The idea is that we keep author/revision if the scanner report told us to copy data from the last analysis. In that case, it means that the SCM information was still present but since the file wasn't changed, the scanner instructed the CE to reuse the data from the DB. In all other cases, we clear the author and revision.